device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
+                  this._d2cLink.on('error', this._d2cErrorListener);
+                  this._d2cLink.send(amqpMessage, handleResult('AMQP Transport: Could not send', sendCallback));
+                }
+              });
+            } else {
+              debug('using existing d2c link');
+              this._d2cLink.send(amqpMessage, handleResult('AMQP Transport: Could not send', sendCallback));
+            }
+          },
+          updateSharedAccessSignature: (sharedAccessSignature, updateSasCallback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_06_010: [If the AMQP connection is established, the `updateSharedAccessSignature` method shall call the amqp transport `putToken` method with the first parameter `audience`, created from the `sr` of the shared access signature, the actual shared access signature, and a callback.]*/
+            this._amqp.putToken(SharedAccessSignature.parse(sharedAccessSignature, ['sr', 'sig', 'se']).sr, sharedAccessSignature, (err) => {
+              if (err) {
+                this._amqp.disconnect(() => {
+                  updateSasCallback(getTranslatedError(err, 'AMQP Transport: Could not authorize with puttoken'));
+                });
+              } else {
+                /*Codes_SRS_NODE_DEVICE_AMQP_06_011: [The `updateSharedAccessSignature` method shall call the `done` callback with a null error object and a SharedAccessSignatureUpdated object as a result, indicating the client does NOT need to reestablish the transport connection.]*/
+                updateSasCallback(null, new results.SharedAccessSignatureUpdated(false));
+              }
+            });
+          },
+          getTwinReceiver: (callback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_026: [** The `getTwinReceiver` method shall call the `done` callback with a `null` error argument and the `AmqpTwinClient` instance currently in use.]*/
+            callback(null, this._twinClient);
+          },
+          sendTwinRequest: (method, resource, properties, body, sendTwinRequestCallback) => {
+            this._twinClient.sendTwinRequest(method, resource, properties, body, sendTwinRequestCallback);
+          },
+          attachC2DLink: (callback) => {
+            debug('attaching C2D link');
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_030: [The `Amqp` object shall attach the C2D `ReceiverLink` object if necessary to start receiving messages.]*/
+            this._amqp.attachReceiverLink(this._c2dEndpoint, null, (err, receiverLink) => {
+              if (err) {
+                debug('error creating a C2D link: ' + err.toString());
+                handleResult('AMQP Transport: Could not attach link', callback)(err);
+              } else {
+                debug('C2D link created and attached successfully');
+                this._c2dLink = receiverLink;
+                this._c2dLink.on('error', this._c2dErrorListener);
+                this._c2dLink.on('message', this._c2dMessageListener);
```

Comment:
I _think_ there's the possibility of attaching the same listener multiple times here, depending on how this._amqp.attachReceiverLink acts.  If that function returns an existing object if it's called twice, then these handlers get added to that object twice.

Reply:
the uniqueness of this link is left as an exercise for the caller for now. this is going to be refactored a bit once I get to the client logic, but for now I believe this works.

---
In reply to: [141132830](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141132830) [](ancestors = 141132830)

Reply:
the uniqueness of this link is left as an exercise for the caller for now. this is going to be refactored a bit once I get to the client logic, but for now I believe this works.

---
In reply to: [141132830](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141132830) [](ancestors = 141132830)